library/Faktory/Job.hs

```diff
@@ -35,26 +36,50 @@ data Job arg = Job
   deriving Generic
 
 perform :: (HasCallStack, ToJSON arg) => Client -> Queue -> arg -> IO JobId
-perform client queue arg = do
-  job <- newJob queue arg
-  jobJid job <$ pushJob client job
+perform = performWith id
 
+-- | Perform the Job at the given time
 performAt
-  :: (HasCallStack, ToJSON arg) => Client -> Queue -> UTCTime -> arg -> IO JobId
-performAt client queue time arg = do
-  job <- newJob queue arg
-  jobJid job <$ pushJob client job { jobAt = Just time }
+  :: (HasCallStack, ToJSON arg) => UTCTime -> Client -> Queue -> arg -> IO JobId
+performAt time = performWith $ \job -> job { jobAt = Just time }
 
+-- | Perform the Job in some time, given as a @'NominalDiffTime'@
 performIn
   :: (HasCallStack, ToJSON arg)
-  => Client
+  => NominalDiffTime
+  -> Client
   -> Queue
-  -> NominalDiffTime
   -> arg
   -> IO JobId
-performIn client queue diff arg = do
+performIn diff = performWithM $ \job -> do
   time <- addUTCTime diff <$> getCurrentTime
-  performAt client queue time arg
+  pure job { jobAt = Just time }
+
+-- | Perform the Job with 0 retries
+performOnce :: (HasCallStack, ToJSON arg) => Client -> Queue -> arg -> IO JobId
+performOnce = performWith $ \job -> job { jobRetry = 0 }
+
+performWith
+  :: (HasCallStack, ToJSON arg)
+  => (Job arg -> Job arg)
+  -> Client
+  -> Queue
+  -> arg
+  -> IO JobId
+performWith f client queue arg = do
+  job <- f <$> newJob queue arg
+  jobJid job <$ pushJob client job
+
+performWithM
+  :: (HasCallStack, ToJSON arg)
+  => (Job arg -> IO (Job arg))
+  -> Client
+  -> Queue
+  -> arg
+  -> IO JobId
+performWithM f client queue arg = do
+  job <- f =<< newJob queue arg
+  jobJid job <$ pushJob client job
```

Comment:
I'm fine with this, but if you really wanna go wild, you can use the monoidal options pattern:

```haskell
perform mempty "queue" SomeJob
perform once "queue" SomeJob
perform (once <> at someTime) "queue" SomeJob
perform (once <> in_ 10) "queue" SomeJob
perform (withRetries 3 <> in_ 10) "queue" SomeJob
```

i.e.
```haskell
-- Simplified, cuz i did this real fast
type Queue = String

data Job a = Job
  { jobRetry :: !Int
  , jobQueue :: !String
  , jobAt :: !(Maybe UTCTime)
  , jobArgs :: !a
  } deriving (Eq, Show, Ord)

data JobUpdate
  = SetRetries Int
  | SetAt UTCTime
  | SetIn NominalDiffTime
  deriving (Eq, Show, Ord)

newtype JobOptions = JobOptions [JobUpdate]
  deriving newtype (Eq, Show, Ord, Semigroup, Monoid)

perform :: JobOptions -> Queue -> a -> IO (Job a)
perform options q a = applyOptions options $ newJob q a

newJob :: Queue -> a -> Job a
newJob q a = Job
  { jobRetry = 25
  , jobQueue = q
  , jobAt = Nothing
  , jobArgs = a
  }

applyOptions :: JobOptions -> Job a -> IO (Job a)
applyOptions (JobOptions patches) = go patches
 where
  go [] job = pure job
  go (x:xs) job =
    case x of
      SetRetries n ->
        go xs $ job {jobRetry = n}
      SetAt t ->
        go xs $ job {jobAt = Just t}
      SetIn i -> do
        t <- getCurrentTime
        go xs $ job {jobAt = Just $ addUTCTime i t}

withRetries :: Int -> JobOptions
withRetries n = JobOptions [SetRetries n]

once :: JobOptions
once = withRetries 0

at :: UTCTime -> JobOptions
at t = JobOptions [SetAt t]

in_ :: NominalDiffTime -> JobOptions
in_ i = JobOptions [SetIn i]
```

Note that this doesn't prevent you from setting the same options more than once, though you can do that with fancier types. Probably not worth the trouble

Reply:
I like this. I like that it avoids `performWithM`. And I like that it's just one `perform` function, having to supply `mempty` in the simplest case seems reasonable to me.